v2/cf-nodejs-logging-support/src/lib/logger/source-utils.ts

```diff
@@ -0,0 +1,114 @@
+    private requestAccessor: RequestAccessor = RequestAccessor.getInstance();
+    private responseAccessor = ResponseAccessor.getInstance();
+
+    private constructor() { }
+
+    public static getInstance(): SourceUtils {
+        if (!SourceUtils.instance) {
+            SourceUtils.instance = new SourceUtils();
+        }
+
+        return SourceUtils.instance;
+    }
+
+    getFieldValue(source: Source, record: any): string | undefined {
+        switch (source.type) {
+            case "static":
+                return source.value;
+            case "env":
+                if (source.path) {
+                    return NestedVarResolver.resolveNestedVariable(process.env, source.path);
+                }
+                return process.env[source.name!];
+            case "config-field":
+                return record[source.name!];
+            default:
+                return undefined;
+        }
+    }
+
+    getReqFieldValue(source: Source, record: any, req: any, res: any): string | undefined {
+        switch (source.type) {
+            case "req-header":
+                return this.requestAccessor.getHeaderField(req, source.name!);
+            case "req-object":
+                return this.requestAccessor.getField(req, source.name!);
+            case "res-header":
+                return this.responseAccessor.getHeaderField(res, source.name!);
+            case "res-object":
+                return this.responseAccessor.getField(res, source.name!);
+            default:
+                return this.getFieldValue(source, record);
+        }
+    }
+
+    // if source is request, then assign to context. If not, then ignore.
+    getContextFieldValue(source: Source, req: any) {
+        switch (source.type) {
+            case "req-header":
+                return this.requestAccessor.getHeaderField(req, source.name!);
+            case "req-object":
+                return this.requestAccessor.getField(req, source.name!);
+            case "uuid":
+                return uuid();
+        }
+    }
+
+    // iterate through sources until one source returns a value 
+    getValueFromSources(record: any, field: ConfigField, origin: origin, req?: any, res?: any) {
+
+        if (origin == "req-log" && (!req || !res)) {
+            throw new Error("Please pass req and res as argument to get value for req-log field.");
+        }
+
+        if (origin == "context" && (!req)) {
+            throw new Error("Please pass req as argument to get value for context field.");
+        }
+
+        field.source = field.source as Source[];
+
+        let sourceIndex = 0;
+        while (!record[field.name]) {
+            sourceIndex = SourceUtils.getInstance().getNextValidSourceIndex(field.source, sourceIndex);
+
+            if (sourceIndex == -1) {
+                return;
+            }
+
+            let source = field.source[sourceIndex];
+
+            record[field.name] = origin == "msg-log" ? this.getFieldValue(source, record) :
```

Comment:
A `get` method should not modify any of its parameters. Please return the value and assign it to `record[field.name]` after calling this method.